google-cloud-tools-plugin/jps-plugin/src/com/google/cloud/tools/intellij/jps/GenRepoInfoFileModuleBuilder.java

```diff
@@ -0,0 +1,169 @@
+
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.jps.ModuleChunk;
+import org.jetbrains.jps.builders.DirtyFilesHolder;
+import org.jetbrains.jps.builders.java.JavaModuleBuildTargetType;
+import org.jetbrains.jps.builders.java.JavaSourceRootDescriptor;
+import org.jetbrains.jps.incremental.BuilderCategory;
+import org.jetbrains.jps.incremental.CompileContext;
+import org.jetbrains.jps.incremental.ModuleBuildTarget;
+import org.jetbrains.jps.incremental.ModuleLevelBuilder;
+import org.jetbrains.jps.incremental.ProjectBuildException;
+import org.jetbrains.jps.model.module.JpsModule;
+
+import java.io.IOException;
+import java.nio.file.Path;
+import java.util.Collections;
+
+/**
+ * Generates source context files for a module using the Cloud SDK.
+ */
+public class GenRepoInfoFileModuleBuilder extends ModuleLevelBuilder {
+
+  private static final Logger LOG =
+      Logger.getInstance("#com.google.cloud.tools.intellij.jps.GenRepoInfoFileModuleBuilder");
+
+  public static final String NAME = "Stackdriver source context generator";
+  private GenRepoInfoFileActionFactory actionFactory = new GenRepoInfoFileActionFactory();
+
+  public GenRepoInfoFileModuleBuilder() {
+    super(BuilderCategory.CLASS_POST_PROCESSOR);
+  }
+
+  @VisibleForTesting
+  GenRepoInfoFileModuleBuilder(GenRepoInfoFileActionFactory actionFactory) {
+    this();
+    this.actionFactory = actionFactory;
+  }
+
+  /**
+   * Generates source context files for each module in {@code chunk}.
+   *
+   * <p>Generation is parameterized by the Stackdriver facet associated to the module.
+   *
+   * <p>Files are added to the same location as the project's compiled .class files.
+   */
+  @Override
+  public ExitCode build(CompileContext context, ModuleChunk chunk,
+      DirtyFilesHolder<JavaSourceRootDescriptor, ModuleBuildTarget> dirtyFilesHolder,
+      OutputConsumer outputConsumer) throws ProjectBuildException, IOException {
+    for (JpsModule jpsModule : chunk.getModules()) {
+      JpsStackdriverModuleExtension extension = jpsModule.getContainer().getChild(
+          JpsStackdriverModuleExtensionImpl.ROLE);
+
+      if (!extension.isGenerateSourceContext()) {
+        continue;
+      }
+
+      if (jpsModule.getSourceRoots().isEmpty()) {
+        LOG.warn("Module " + jpsModule.getName() + " contains no source roots. Moving on to the "
+            + "next module for source generation.");
+        continue;
+      }
+
+      if (extension.getCloudSdkPath() == null) {
+        LOG.warn("No Cloud SDK path specified. Skipping source context generation.");
+        // Cloud SDK path is a singleton for a project, so if it doesn't exist on the first module,
+        // we can return right away.
+        return ExitCode.NOTHING_DONE;
+      }
+
+      if (extension.getModuleSourceDirectory() == null) {
+        // Possible to happen on Windows, unlikely in Linux.
+        LOG.warn("Invalid module source directory. Check for special characters <>:\"|?*.");
```

Comment:
By the way, this null only happens if a path is invalid by having special characters. If it is a wellformed path without a Cloud SDK installation, this doesn't trigger (i.e., the return isn't null).

I must be missing something obvious, but why is it unintuitive? We're doing something similar for the Cloud SDK path.

Even if we didn't save an invalid Cloud SDK path, getModuleSourceDirectory() default return would still be null.

Using validators here, as opposed to a null return, is also not possible, since they're currently built into the IJ plugin. We could be on to something if the validators were implemented in appengine-plugins-core, which jps-plugin has access to.

What's your change suggestion here?